src/modules/form/util/expressions/formula.ts

```diff
@@ -0,0 +1,70 @@
+import ExpressionEvaluationError from './ExpressionEvaluationError';
+import formulaFunctions from './formulaFunctions';
+import { parseExpression } from './parser';
+import { parseNumber } from '@/utils/numbers';
+
+const ensureNumeric = (value: any): number => {
+  if (parseNumber(value) === undefined)
+    throw new ExpressionEvaluationError('Non numeric result');
+  return value;
+};
+
+type EvalContext = Map<string, any>;
+const evaluate = (ast: any, context: EvalContext): number | undefined => {
```

Comment:
This is cool and very flexible!
Can you explain a little bit how the variables get evaluated? I'm getting a little lost at that part. In your example `end_date - start_date` where do those values get read from?

Reply:
Sure! So the formula expression is processed in two stages. It is pretty much the same as how we handle `sumQuestions`, I just copied that plumbing.

The first stage simply collects all the variable names referenced by the expression statement (evaluator encounters `ast.type === "Identifier"`). Then each variable referenced by the formula is aded as a dependency. That's this code in formUtils:
```js
for (const id of collectExpressionReferences(av.formula)) { addDependency(id); }
// for "end_date - start_date",  this is: addDependency("end_date"); addDependency("start_date")
```

In the second stage the values are read and operations are performed to return the result. We construct a new context map to provide the current values to the evaluator. This includes all variables (`Object.keys(values)`) which is probably too broad and could be narrowed. The `evaluateFormula` function only uses the values referenced in the expression from the context. Note, instead of constructing a map, we could use the values object directly; the map provides a more safety and control in exchange for overhead.
```js
const context = new Map();
for (const key of Object.keys(values)) {
  const value = numericValueForFormValue(values[key]);
  if (!isNil(value)) context.set(key, value);
}
return evaluateFormula(av.formula, context);
// context would be new Map([['end_date', 120], ['start_date', 60], ... other keys in the values object])
```

Reply:
👍  awesome.